subprojects/problems/src/main/java/org/gradle/api/problems/internal/DefaultProblemBuilder.java

```diff
@@ -139,15 +147,21 @@ public Problem build() {
             throw new IllegalStateException("Problem is not documented: " + message);
         }
 
-        if(!noLocation && (path == null || line == null)) {
-            throw new IllegalStateException("Problem has no location: " + message);
+        if (!noLocation) {
+            if (path == null) {
+                throw new IllegalStateException("Problem location path is not set: " + message);
+            }
+            if (line == null) {
+                throw new IllegalStateException("Problem location line is not set: " + message);
+            }
```

Comment:
🤔  I think this might be more useful done this way. Feel free to reject this if you don't like it.